source/MaterialXGenShader/Shader.h

```diff
@@ -31,291 +23,77 @@ using ShaderPtr = shared_ptr<class Shader>;
-    const VariableBlock& getAppDataBlock() const { return _appData; }
-
-    /// Return the block of output variables.
-    const VariableBlock& getOutputBlock() const { return _outputs; }
-
-    /// Start a new scope in the shader, using the given bracket type
-    virtual void beginScope(Brackets brackets = Brackets::BRACES);
-
-    /// End the current scope in the shader
-    virtual void endScope(bool semicolon = false, bool newline = true);
-
-    /// Start a new line in the shader
-    virtual void beginLine();
-
-    /// End the current line in the shader
-    virtual void endLine(bool semicolon = true);
-
-    /// Add a newline character to the shader
-    virtual void newLine();
-
-    /// Add a string to the shader
-    virtual void addStr(const string& str);
+    /// Return a stage by index.
+    ShaderStage& getStage(size_t index);
 
-    /// Add a single line of code to the shader,
-    /// optionally appening a semi-colon
-    virtual void addLine(const string& str, bool semicolon = true);
+    /// Return a stage by index.
+    const ShaderStage& getStage(size_t index) const;
 
-    /// Add a block of code to the shader
-    virtual void addBlock(const string& str, ShaderGenerator& shadergen);
+    /// Return a stage by name.
+    ShaderStage& getStage(const string& name);
 
-    /// Add the function definition for a node
-    virtual void addFunctionDefinition(ShaderNode* node, ShaderGenerator& shadergen);
+    /// Return a stage by name.
+    const ShaderStage& getStage(const string& name) const;
 
-    /// Add the function call for a node
-    virtual void addFunctionCall(ShaderNode* node, const GenContext& context, ShaderGenerator& shadergen);
-
-    /// Add the contents of an include file
-    /// Making sure it is only included once
-    /// for a shader stage
-    virtual void addInclude(const string& file, ShaderGenerator& shadergen);
-
-    /// Add a single line of code comment to the shader
-    virtual void addComment(const string& str);
-
-    /// Add a value to the shader
-    template<typename T>
-    void addValue(const T& value)
+    /// Return true if the shader has a given named attribute.
+    bool hasAttribute(const string& attrib) const
     {
-        std::stringstream str;
-        str << value;
-        _activeStage->code += str.str();
+        return _attributeMap.count(attrib) != 0;
     }
 
-    /// Return the shader name
-    const string& getName() const { return _name; }
+    /// Return the value for a named attribute,
+    /// or nullptr if no such attribute is found.
+    ValuePtr getAttribute(const string& attrib) const
+    {
+        auto it = _attributeMap.find(attrib);
+        return it != _attributeMap.end() ? it->second : nullptr;
+    }
 
-    /// Return the active shader graph.
-    ShaderGraph* getGraph() const { return _graphStack.back(); }
+    /// Set a value attribute on the shader.
+    void setAttribute(const string& attrib, ValuePtr value)
```

Comment:
Added general attributes on Shaders. For storing any extra attributes needed, e.g. the transparency flag on HW shaders.